core/state_metrics/impl/state_metrics_impl.cpp

```diff
@@ -0,0 +1,201 @@
+    BOOST_ASSERT(state_api_);
+    BOOST_ASSERT(registry);
+    BOOST_ASSERT(hasher);
+    auto validator_id = primitives::decodeSs58(validator_address, *hasher);
+    if (not validator_id) {
+      const auto error_message =
+          std::string("Failed to decode validator address: ")
+          + validator_id.error().message();
+      throw std::runtime_error(error_message);
+    }
+    validator_id_ = std::move(validator_id.value());
+    constexpr auto era_points_metric = "era_points";
+    registry->registerGaugeFamily(
+        era_points_metric,
+        "The number of reward points for the current era for the validator",
+        {{"validator_address", validator_address}});
+    era_points_gauge_ = registry->registerGaugeMetric(era_points_metric);
+    BOOST_ASSERT(era_points_gauge_);
+
+    const auto staking_prefix =
+        crypto::make_twox128(kagome::common::Buffer::fromString("Staking"));
+    const auto active_era_prefix =
+        crypto::make_twox128(kagome::common::Buffer::fromString("ActiveEra"));
+    active_era_storage_key_.insert(active_era_storage_key_.end(),
+                                   staking_prefix.begin(),
+                                   staking_prefix.end());
+    active_era_storage_key_.insert(active_era_storage_key_.end(),
+                                   active_era_prefix.begin(),
+                                   active_era_prefix.end());
+
+    const auto reward_points_prefix = crypto::make_twox128(
+        kagome::common::Buffer::fromString("ErasRewardPoints"));
+    reward_points_storage_key_basis_.insert(
+        reward_points_storage_key_basis_.end(),
+        staking_prefix.begin(),
+        staking_prefix.end());
+    reward_points_storage_key_basis_.insert(
+        reward_points_storage_key_basis_.end(),
+        reward_points_prefix.begin(),
+        reward_points_prefix.end());
+
+    era_points_thread_ = std::thread([this] {
+      for (;;) {
+        scheduler_->schedule([WEAK_SELF] {
+          WEAK_LOCK(self);
+          self->updateEraPoints();
+        });
+        for (auto i = 0; i < SET_ERA_POINTS_PERIOD; ++i) {
+          std::this_thread::sleep_for(std::chrono::seconds(1));
+          if (stop_signal_received_.load()) {
+            return;
+          }
+        }
+      }
+    });
+  }
+
+  outcome::result<std::shared_ptr<StateMetricsImpl>> StateMetricsImpl::create(
+      const application::AppConfiguration &app_config,
+      std::shared_ptr<libp2p::basic::Scheduler> scheduler,
+      std::shared_ptr<api::StateApi> state_api,
+      std::shared_ptr<metrics::Registry> registry,
+      std::shared_ptr<crypto::Hasher> hasher) {
+    if (auto validator_address = app_config.getValidatorAddress()) {
+      return std::make_shared<StateMetricsImpl>(
+          std::move(validator_address.value()),
+          std::move(scheduler),
+          std::move(state_api),
+          std::move(registry),
+          std::move(hasher));
+    }
+    return nullptr;
```

Comment:
I strongly disagree with the case of sptr being initialized with nullptr. 
A possible solution could be wrapping it with an optional.

Reply:
I agree. Signature of the function could be just `optional<shared_ptr<...>>`